### ISSUE TITLE
False positive of `InterpolationCheck` for strings inside strings

### ISSUE THREAD

**Issue by AlexWayfer** [rubocop-hq/rubocop#8416]
Hello.

I have a code, tests, like this:

```ruby
let(:expected_words) do
  [
    'module FooBar',
    'FB::Application.logger.info "#{mail.log_message} [#{index}/#{count}]..."',
    'File.join(FB::Application.config[:tmp_dir], "mailing_#{object_id}")'
  ]
end
```

So, there is a String in single quotes which contains a code with a String in double quotes and interpolation.

--------

## Expected behavior

No offenses.

## Actual behavior

```
Lint/InterpolationCheck: Interpolation in single quoted string detected. Use double quoted strings if you need interpolation.
```

## Steps to reproduce the problem

Code is above.

## RuboCop version

```
0.88.0 (using Parser 2.7.1.4, rubocop-ast 0.2.0, running on ruby 2.6.6 x86_64-linux)
```

**Comment by marcandre** [comment]
My understanding is that the whole point of the cop is to detect this (which looks like a bug but isn't)

You may want to either use `rubocop:disable Lint/InterpolationCheck`,  exclude your spec folder for that cop or disable this cop entirely.

I note that this cop is considered Safe but with Unsafe correction. I'm marking it as Unsafe altogether because the diagnostic itself is not assured.

**Comment by AlexWayfer** [comment]
> My understanding is that the whole point of the cop is to detect this (which looks like a bug but isn't)

We can just check that an interpolation in double quotes inside single quotes, can't we?

**Comment by AlexWayfer** [comment]
> > My understanding is that the whole point of the cop is to detect this (which looks like a bug but isn't)
> 
> We can just check that an interpolation in double quotes inside single quotes, can't we?

@marcandre, what do you think?

**Comment by marcandre** [comment]
> We can just check that an interpolation in double quotes inside single quotes, can't we?

IIUC, you would like that this cop considers `'..."#{...}"...'` ok but `'...#{...}...'` not ok. I think both of theses patterns are highly unusual and it makes sense for RuboCop to question them both. There's nothing wrong with disabling a Cop here and there for exceptional cases.

If there was a common usecase that seemed right that might be reconsidered, but at first sight your test seems very unusual and looks [extremely sensitive to the implementation](https://testing.googleblog.com/2015/01/testing-on-toilet-change-detector-tests.html).

**Comment by AlexWayfer** [comment]
> If there was a common usecase that seemed right

What is the "common usecase" for you?

**Comment by marcandre** [comment]
I'm not sure I understand your question. A "common use-case" would be an example of code that seems legitimate yet contradicts a cop.

**Comment by MatzFan** [comment]
Just to add my 2 cents here, as I've encountered this false positive more than once in my code.

My personal opinion is that Rubocop should not try and second guess the user's intention here, especially as the user's text editor/IDE will almost certainly correctly highlight the lack of interpolation in such strings. E.g. Sublime Text's treatment of the OP's code:

![st](https://github.com/user-attachments/assets/9f3a0e5c-f44c-49b3-a750-a504087e7791)

String interpolation example for comparison:

![st2](https://github.com/user-attachments/assets/c7c1b886-4773-4597-9fab-1e85a8dc1dd2)

IMHO this cop is redundant for all but the exception case of a user who uses a text editor without Ruby syntax highlighting. Heck, even GitHub flavored markdown shows the difference. Compare this with the OP's code block:

```ruby
let(:expected_words) do
  [
    'module FooBar',
    "FB::Application.logger.info #{mail.log_message} [#{index}/#{count}]...",
    'File.join(FB::Application.config[:tmp_dir], "mailing_#{object_id}")'
  ]
end
```
My vote would be to ditch this cop.
